### PR TITLE
feat: tree-shake deterministic dynamic imports

### LINKED ISSUE #3417: Using lazy imports seems to cause a size increase in the eagerly imported entry point.
- Rollup Version: 1.32.0
- Operating System (or Browser): Windows 10
- Node Version (if applicable): v12.4.0
- Link to reproduction (IMPORTANT, read below):

  - `git clone https://github.com/filipesilva/rollup-aio-lazy`
  - `cd rollup-aio-lazy` 
  - `yarn`
  - `yarn rollup --dir dist-full`
  - Open `src/app/custom-elements/element-registry.js` and comment out this part:
  ```js
      // Comment these out to remove lazy routes.
      // Start here:
      {
          selector: 'aio-announcement-bar',
          loadChildren: () => import('./announcement-bar/announcement-bar.module').then(m => m.AnnouncementBarModule)
      },
      {
          selector: 'aio-api-list',
          loadChildren: () => import('./api/api-list.module').then(m => m.ApiListModule)
      },
      {
          selector: 'aio-contributor-list',
          loadChildren: () => import('./contributor/contributor-list.module').then(m => m.ContributorListModule)
      },
      {
          selector: 'aio-file-not-found-search',
          loadChildren: () => import('./search/file-not-found-search.module').then(m => m.FileNotFoundSearchModule)
      },
      {
          selector: 'aio-resource-list',
          loadChildren: () => import('./resource/resource-list.module').then(m => m.ResourceListModule)
      },
      {
          selector: 'aio-toc',
          loadChildren: () => import('./toc/toc.module').then(m => m.TocModule)
      },
      {
          selector: 'code-example',
          loadChildren: () => import('./code/code-example.module').then(m => m.CodeExampleModule)
      },
      {
          selector: 'code-tabs',
          loadChildren: () => import('./code/code-tabs.module').then(m => m.CodeTabsModule)
      },
      {
          selector: 'live-example',
          loadChildren: () => import('./live-example/live-example.module').then(m => m.LiveExampleModule)
      }
      // Stop here.
  ```
  - `yarn rollup --dir dist-nolazy`
  - Compare the sizes of the main file in `dist-full` and `dist-nolazy`, see the full version is ~50kb larger:
  ```
  $ ls -la dist-full/main-rollup-91f3bcc1.js
  -rw-r--r-- 1 kamik 197609 1899825 Mar  2 19:40 dist-full/main-rollup-91f3bcc1.js

  $ ls -la dist-nolazy/main-rollup.js
  -rw-r--r-- 1 kamik 197609 1850516 Mar  2 19:42 dist-nolazy/main-rollup.js
  ```
  - If you are using VSCode, you can open those two files and run `File: Compare Active File With...` to see a diff between the two. Some of the extra retained symbols, like `FocusKeyManager` and `A11yModule`, are declared in `main`, unused there, exported, and only used in one of the lazy modules.

### Expected Behavior
I expected that adding lazy imports would not add extra code to the eager code.

### Actual Behavior
Adding lazy imports brought in extra code that should be in the lazy chunks.

### LINKED ISSUE #3447: Tree shaking in dynamic modules
### Feature Use Case

[Repl example](https://rollupjs.org/repl/?version=2.1.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmFzeW5jJTIwZnVuY3Rpb24lMjBtYWluKCklMjAlN0IlNUNuJTVDdGNvbnN0JTIwJTdCJTIwZXhwb3J0MSUyMCU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCcuJTJGdXRpbHMuanMnKSUzQiU1Q24lNUN0ZXhwb3J0MSgpJTNCJTVDbiU3RCU1Q24lNUNubWFpbigpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIydXRpbHMuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZnVuY3Rpb24lMjBleHBvcnQxKCklMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKDEpJTNCJTVDbiU3RCU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBleHBvcnQyKCklMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKDEpJTNCJTVDbiU3RCUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlc20lMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==).

`export2` is unused but appears in the bundle.

### Feature Proposal

Object shape tracking was mentioned in https://github.com/rollup/rollup/issues/3417#issuecomment-594363336.

I'm aware this issue has been discussed in many other issues, but it felt like it didn't have its own issue. Feel free to close if it doesn't need one.

### LINKED ISSUE #4951: Treeshake "static" dynamic import
### Feature Use Case

Treeshake is such a brilliant idea. Nowadays, it powers so many frameworks/libraries to innovate without worries too much about the bundle size.

It's great that you can take some parts from a package, without paying the cost of the entire bundle:

```ts
import { unique } from 'lodash-es'
```

It works great on static imports. However, sometimes when you want to make some part lazy-load, treeshaking stopped to work:

```ts
const { unique } = await import ('lodash-es') // entire `lodash-es` will be bundled
```

###### Workaround

There is a workaround to this, is that you can create a temporary module that does the reexport:

```ts
// ./re-export.js
import { unique } from 'lodash-es'
```

```ts
const { unique } = await import ('./re-export.js') // now you have tree-shaking
```

However, this makes it verbose to use and easy to miss.

### Feature Proposal

Support tree-shaking dynamic imports that are known at build time. Namely, doing object  destructure and directly assigning with `await import()`

###### Valid cases

```ts
const { foo, default: anotherName } = await import('foo')
```

```ts
const { foo } = await import('foo')
const { bar } = await import('foo')
```

```ts
(await import('foo')).foo
```

###### Invalid cases

(Tree shaking is not applicable, to the current behavior.)

```ts
// full bundle
const foo = await import('foo')
```

```ts
// not waiting
const promise = import('foo')
```

```ts
// rest
const { foo, ...bar } = await import('foo')
```

```ts
// imported multiple times, but one of them is not tree shakable
const { foo } = await import('foo')
const foo = await import('foo')
```

----

Even though the "valid cases" are a bit strict, I believe it's still a very common pattern for people to benefit from. 

----

Idea credits: Webpack 5.80 https://twitter.com/TheLarkInn/status/1648739331807510528

### PR DESCRIPTION
**Pull request by antfu**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

resolves #4951

fix #3417
fix #3447

### Description

Full proposal in #4951

This PR introduces a new tree shake option - `deterministicDynamicImports` (enabled by the default preset)

It enabled the tree-shaking for deterministic dynamic imports. The following are valid cases:

```ts
const { foo } = await import('foo')
```

```ts
(await import('foo')).foo
```

```ts
import('bar').then(({ foo }) => {})
```

Idea credits to:
- webpack: https://github.com/webpack/webpack/pull/16995
- Parcel: https://parceljs.org/features/scope-hoisting/#dynamic-imports

### CODE REVIEW

**Review by lukastaegert — COMMENTED** [review]
Great work! I will need some more time for proper review, but from a first glance this looks like a very good direction as it does not add a lot of overhead and touches few places.
Some thoughts: Why did you add a feature flag? Is there something that might be broken by that feature? Otherwise I would advise against it, we already have too many options and I see no benefit of turning it off (the other things that can be turned off all come with some issues if you turn them on). This also means it does not need to be documented etc.
Just make sure that if tree shaking is set to false entirely, then this logic is bypassed as well or at least does not have an effect. Better add a test for this.
Beyond that, I think there is some low-hanging fruit left with side-effect only dynamic imports. I.e. if the return value of `await` is unused because it is an ExpressionStatement, or `then` has no arguments.

**Comment by antfu** [comment]
I wasn't sure if it should be an opt-in feature, or able to opt-out. But yeah I think that make sense and I will removed it :)

**Comment by antfu** [comment]
Now I have implemented the side effect of only imports, which breaks many tests that were previously assumed dynamic imports will always be included. I disabled the tree-shaking for those tests to move on as I think the new behavior is more expected.

File: src/ast/nodes/ImportExpression.ts [hunk 1]
```diff
@@ -296,3 +394,15 @@ const accessedImportGlobals: Record<string, string[]> = {
 	cjs: ['require'],
 	system: ['module']
 };
+
+function getDeterministicObjectDestructure(objectPattern: ObjectPattern): string[] | undefined {
+	const variables: string[] = [];
+
+	for (const property of objectPattern.properties) {
+		if (property.type === 'RestElement' || property.computed || property.key.type !== 'Identifier')
```
Comment (antfu) [inline src/ast/nodes/ImportExpression.ts:402]: I am not sure how to cover the `.computed` case, but I think it's better to have the check to be more defensive.

**Comment by lukastaegert** [comment]
I would prefer not to turn off tree-shaking unless that is explicitly what is tested. If some tests produce empty chunks now and it seems that is not what we want, than rather add a side effect to the files e.g. a `console.log(...)`. As long as Not `function` tests are broken (those should always be thoroughly investigated if broken!), you can mass update snapshots via `npm run test:update-snapshots`. This can also help to illustrate the impact of the feature on a code base that was not specifically designed for the feature.
Disabling tree-shaking is an edge case while tree-shaking is also something we recommend everyone to keep switched on. Thus it should also be enabled in tests.

**Comment by antfu** [comment]
Yeah you are right. Actually it's due to I didn't handle side-effects right. It should be fixed now.

File: src/ast/nodes/ImportExpression.ts [hunk 2]
```diff
@@ -45,6 +54,95 @@ export default class ImportExpression extends NodeBase {
...
+	 * - `const { foo } = await import('bar')`.
+	 * - `(await import('bar')).foo`
+	 * - `import('bar').then(({ foo }) => {})`
+	 *
+	 * Returns empty array if it's side-effect only import.
+	 * Returns undefined if it's not fully deterministic.
+	 */
+	getDeterministicImportedNames(): readonly string[] | undefined {
+		// side-effect only
+		if (this.parent.type === 'ExpressionStatement') {
```
Comment (lukastaegert) [inline src/ast/nodes/ImportExpression.ts:69]: This looks like a three-case switch-statement over `this.parent.type`, which would also avoid the repeated property accesses.

Alternatively, a better approach that is also used in other places and which solves TypeScript issues would be to use instanceof:

```ts
const {parent} = this;
if (parent instanceof ExpressionStatement) {
  return EMPTY_ARRAY;
}
if (parent instanceof AwaitExpression) {
  // Now the type is narrowed
  // It is always a good idea to avoid repeated property accesses for both performance and typing reasons
  const awaitParent = parent.parent;
  if (awaitParent instanceof ExpressionStatement) {
    return EMPTY_ARRAY;
  }
  if (awaitParent instanceof VariableDeclarator) {
    // Here the type-narrowing means we have id correctly typed
    if (!(awaitParent.id instanceof ObjectPattern)) return;
    ...
  }
}
```

One thing to look out for is if this creates cyclic dependencies. In that case, you need to check `type` after all. But otherwise, it should be marginally more performant and always has the right types.

File: src/ast/nodes/ImportExpression.ts [hunk 3]
```diff
@@ -45,6 +54,95 @@ export default class ImportExpression extends NodeBase {
...
+
+			// Case 1: const { foo } = await import('bar')
+			if (awaitExpression.parent?.type === 'VariableDeclarator') {
+				const variableDeclarator = awaitExpression.parent as VariableDeclarator;
+				if (variableDeclarator.id?.type !== 'ObjectPattern') return;
+
+				return getDeterministicObjectDestructure(variableDeclarator.id as ObjectPattern);
+			}
+			// Case 2: (await import('bar')).foo
+			else {
```
Comment (lukastaegert) [inline src/ast/nodes/ImportExpression.ts:89]: You can unwrap the `else` case after a `return` here.

File: src/ast/nodes/ImportExpression.ts [hunk 4]
```diff
@@ -45,6 +54,95 @@ export default class ImportExpression extends NodeBase {
...
+				if (awaitExpression.parent?.type !== 'MemberExpression') return;
+
+				const memberExpression = awaitExpression.parent as MemberExpression;
+				if (memberExpression.computed || memberExpression.property.type !== 'Identifier') return;
+
+				return [(memberExpression.property as Identifier).name];
+			}
+		}
+		// Case 3: import('bar').then(({ foo }) => {})
+		else if (this.parent?.type === 'MemberExpression') {
```
Comment (lukastaegert) [inline src/ast/nodes/ImportExpression.ts:99]: Again, you can unwrap the `else` case here.

File: src/ast/nodes/ImportExpression.ts [hunk 5]
```diff
@@ -45,6 +54,95 @@ export default class ImportExpression extends NodeBase {
...
+			const memberName = (memberExpression.property as Identifier).name;
+			const callExpression = memberExpression.parent as CallExpression;
+
+			// side-effect only, when only chaining .catch or .finally
+			if (
+				callExpression.parent.type === 'ExpressionStatement' &&
+				['catch', 'finally'].includes(memberName)
+			) {
+				return EMPTY_ARRAY;
+			} else if (memberName !== 'then') {
```
Comment (lukastaegert) [inline src/ast/nodes/ImportExpression.ts:116]: Again this can be unwrapped.

File: src/ast/nodes/ImportExpression.ts (hunk 1 above)
Comment (lukastaegert) [reply src/ast/nodes/ImportExpression.ts:402]: Maybe something like https://rollup-46n0bvzzx-rollup-js.vercel.app/repl/?pr=4952&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTdCJTVCZ2xvYmFsVGhpcy51bmtub3duJTVEJTNBJTIwZm9vJTJDJTIwYmFyJTdEJTIwJTNEJTIwYXdhaXQlMjBpbXBvcnQoJy4lMkZkZXAuanMnKSUzQiU1Q25jb25zb2xlLmxvZyhmb28pJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjIlMkYlMkYlMjBNdXN0JTIwYmUlMjBpbmNsdWRlZCU1Q25leHBvcnQlMjBjb25zdCUyMGElMjAlM0QlMjAnc29tZXRoaW5nJyUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSUyQyUyMm5hbWUlMjIlM0ElMjJkZXAuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0ElMjJzbWFsbGVzdCUyMiU3RCU3RA==

**Review by lukastaegert — CHANGES_REQUESTED** [review]
Thanks, this looks really great! Please reenable tree-shaking for the given tests (and possibly add side effects if needed), and there were some small suggestions. Looking forward to merging this!

File: src/Module.ts [hunk 6]
```diff
@@ -716,6 +716,27 @@ export default class Module {
...
 
+	includeExportsByNames(names: readonly string[], includeNamespaceMembers: boolean): void {
+		if (!this.isExecuted) {
+			markModuleAndImpureDependenciesAsExecuted(this);
+			this.graph.needsTreeshakingPass = true;
+		}
+
+		for (const name of names) {
+			const variable = this.getVariableForExportName(name)[0];
+			if (variable) {
```
Comment (lukastaegert) [inline src/Module.ts:727]: The case where `variable` is `null` corresponds to the situation where we destructure something that is not actually exported. This is fine—we just get `undefined`—but I think this should be covered in a test. I.e.

```js
const {doesNotExist} = import('./foo.js');
```

File: src/Module.ts [hunk 7]
```diff
@@ -716,6 +716,27 @@ export default class Module {
...
+				variable.deoptimizePath(UNKNOWN_PATH);
+				if (!variable.included) {
+					this.includeVariable(variable);
+				}
+			}
+		}
+
+		if (includeNamespaceMembers) {
```
Comment (lukastaegert) [inline src/Module.ts:735]: Setting this always to `true` is fine, but there are cases where we can actually set this to `false`. Let me explain.
There are two situations where the namespace will contain exports that we do not know at build time:

1. External namespace reexports `export * from 'external'`
2. Synthetic namespaces, i.e. unresolved exports will be picked as properties from an object export.

In both cases, those exports have lower priority than any existing named exports. So if we only import variables that are actually exported in the module, or reexported from another module by name. To check for this, you can verify

```js
this.exports.has(name) || this.reexportDescriptions.has(name);
```

If that is the case for all names, you can skip the "merged namespaces". So why would this be beneficial, and how can you test this? This is relevant in situations where the namespace is rendered as an object. For dynamic imports, this is always the case if the file has additional exports that are not part of the namespace. Then it gets complicated, see this test-case:
https://rollup-ebmohmssk-rollup-js.vercel.app/repl/?pr=4952&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTdCZm9vJTJDeCU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCcuJTJGZGVwMS5qcycpJTVDbmNvbnNvbGUubG9nKGZvbyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZm9vJTIwJTNEJTIwMyUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGJhciUyMCUzRCUyMDQlM0IlNUNuZXhwb3J0JTIwKiUyMGZyb20lMjAnZXh0ZXJuYWwnJTVDbiU1Q25pbXBvcnQoJy4lMkZkZXAyJyklM0IlNUNuaW1wb3J0JTIwJTdCc2hhcmVkJTdEJTIwZnJvbSUyMCcuJTJGc2hhcmVkJyUzQiU1Q25jb25zb2xlLmxvZyhzaGFyZWQpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmRlcDEuanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwJy4lMkZkZXAxJyU1Q25pbXBvcnQlMjAlN0JzaGFyZWQlN0QlMjBmcm9tJTIwJy4lMkZzaGFyZWQnJTNCJTVDbmNvbnNvbGUubG9nKHNoYXJlZCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlMkMlMjJuYW1lJTIyJTNBJTIyZGVwMi5qcyUyMiU3RCUyQyU3QiUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMHNoYXJlZCUyMCUzRCUyMCdmb28nJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMnNoYXJlZC5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlN0QlN0Q=

Here the dynamic import target has itself a dynamic import, and both imported files share a dependency. That dependency is hoisted into the first import target, tainting the namespace. In the example, we restructure `x` that does not exist. If we would not do that, we would not need any of the mergedNamespace helper logic. To see what this would look like, just remove the `export * from 'external'` logic.

Not sure if you want to add this optimization, but at least it is not costly.

Comment (lukastaegert) [reply src/Module.ts:735]: Ah, I just noticed a much simpler test case is to use `inlineDynamicImports`: https://rollup-ebmohmssk-rollup-js.vercel.app/repl/?pr=4952&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwJTdCZm9vJTJDeCU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCcuJTJGZGVwMS5qcycpJTVDbmNvbnNvbGUubG9nKGZvbyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwZm9vJTIwJTNEJTIwMyUzQiU1Q25leHBvcnQlMjBjb25zdCUyMGJhciUyMCUzRCUyMDQlM0IlNUNuZXhwb3J0JTIwKiUyMGZyb20lMjAnZXh0ZXJuYWwnJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMmRlcDEuanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJpbmxpbmVEeW5hbWljSW1wb3J0cyUyMiUzQXRydWUlN0QlN0QlN0Q=

File: src/ast/nodes/ImportExpression.ts [hunk 8]
```diff
@@ -45,6 +55,100 @@ export default class ImportExpression extends NodeBase {
...
+				const declaration = parent2.id;
+				return declaration instanceof ObjectPattern
+					? getDeterministicObjectDestructure(declaration)
+					: undefined;
+			}
+
+			// Case 2: (await import('bar')).foo
+			if (parent2 instanceof MemberExpression) {
+				const id = parent2.property;
+				if (!parent2.computed && id instanceof Identifier) {
```
Comment (lukastaegert) [inline src/ast/nodes/ImportExpression.ts:95]: It appears this `computed` case is the only missing test case, everything else is covered now. Good work!

Probably something like `await import('bar')[foo]`...

**Review by lukastaegert — COMMENTED** [review]
Just some final comments and a possible optimization I stumbled upon.

File: test/form/samples/treeshake-deterministic-dynamic-import/main.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  // side-effect only
+  import('./effect-1.js')
+  await import('./effect-2.js')
+  import('./effect-3.js').then(function() { })
+  import('./effect-4.js').then()
+  import('./effect-5.js').catch(() => {})
+  import('./effect-6.js').finally(() => {})
+
+  // bail out
+  const { named1 } = await import('./bail-1.js');
```
Comment (lukastaegert) [inline test/form/samples/treeshake-deterministic-dynamic-import/main.js:21]: One last thing that occurred to me would be to double the "bail out" cases as function tests. `form` tests are basically snapshot tests, which means they need to be reviewed thoroughly by a human when anything changes and problems are easily overlooked. Unfortunately, they are the only way to test tree-shaking. `function` tests are better in so far as they actually execute the code. They also get `assert` as a global variable so you can make inline assertions. It could look like this:

```js
// main.js
export const allExports = Promise.all([
  (async () => {
    const value = (await import('./bail-3.js'))[foo];
    assert.strictEqual(value, '...');
  })(),
  ...
]);

// _config.js
...
async exports({allExports}) {
  // to ensure async errors are caught
  await allExports;
},
context: {
  foo: '...'
}
```

File: src/ast/nodes/ImportExpression.ts (hunk 8 above)
Comment (antfu) [reply src/ast/nodes/ImportExpression.ts:95]: We have the case already and the computed has already been covered - the missing coverage is the `id instanceof Identifier` - as when it's not computed, it seems it will only be able to be Identifier, we don't have such `(await import('bar')).0`. I think having this check would be good to be defensive but I guess we could also do a force type casting.

File: test/form/samples/treeshake-deterministic-dynamic-import/main.js (hunk 9 above)
Comment (antfu) [reply test/form/samples/treeshake-deterministic-dynamic-import/main.js:25]: I found that it would be hard to assert what had been shaken or not in the runtime - so I go with another approach to assert for the generated code with certain marks. Do you think that's a good approach?

Comment (lukastaegert) [reply test/form/samples/treeshake-deterministic-dynamic-import/main.js:25]: No, I am not a fan of explicit code assertions. After all, the generated code may change drastically with new Rollup versions while observable effects must not change. Then stick with form tests. But we are bailing out for a reasons. We are not keeping the exports because they are destructured, but because they might actually be used. And asserting their values are right after destructuring is a very simple and effective way of using them. This will also make the test much less artificial.
In `function` tests, you can inject globals via `context`, can await exported promises via `Exports` and have `assert` as a global. This is usually enough material to write tests that are close to what actual users might do and which will never need to be changed if Rollup becomes more advanced.

Comment (lukastaegert) [reply test/form/samples/treeshake-deterministic-dynamic-import/main.js:25]: Of course you cannot test what is removed with such a test. For that, you NEED the snapshot/chunking-form test. But the more critical error is if something is removed that we need to keep because that leads to broken code. And for that we have `function` tests.

Comment (lukastaegert) [reply test/form/samples/treeshake-deterministic-dynamic-import/main.js:25]: I.e. keep the current test as it is, but add a second function test that just asserts that nothing is removed that is actually used.

**Review by lukastaegert — APPROVED** [review]
Perfect, thanks a lot!

File: src/ast/nodes/ImportExpression.ts [hunk 10]
```diff
@@ -45,6 +55,100 @@ export default class ImportExpression extends NodeBase {
 		this.source.bind();
 	}
 
+	/**
+	 * Get imported variables for deterministic usage, valid cases are:
+	 *
+	 * - `const { foo } = await import('bar')`.
+	 * - `(await import('bar')).foo`
+	 * - `import('bar').then(({ foo }) => {})`
```
Comment (dasa) [inline src/ast/nodes/ImportExpression.ts:63]: Should this be doc'ed here too? https://rollupjs.org/introduction/#tree-shaking

Comment (lukastaegert) [reply src/ast/nodes/ImportExpression.ts:63]: Any PR for the docs is welcome! I guess something similar to what Parcel has does not hurt.

Comment (dasa) [reply src/ast/nodes/ImportExpression.ts:63]: Is this what you're referring to? https://parceljs.org/features/scope-hoisting/#dynamic-imports

Comment (lukastaegert) [reply src/ast/nodes/ImportExpression.ts:63]: 👍
